{"cells":[{"cell_type":"markdown","metadata":{},"source":["# Trabalho de Bioinformática\n","- Ana Cristina Silva de Oliveira, 11965630\n","- Fernando Henrique Paes Generich, 11795342\n","- Vítor Amorim Fróis, 12543440"]},{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":9821,"status":"ok","timestamp":1732148125524,"user":{"displayName":"Fernando .Random. Henrique Paes Generich","userId":"13317897141628915437"},"user_tz":180},"id":"qokabCeg_taP","outputId":"d57ab6b0-4a06-4dfc-d5a7-f9a4186c39e7"},"outputs":[],"source":["import pandas as pd\n","from Bio import Entrez, SeqIO, Seq\n","from tqdm import tqdm"]},{"cell_type":"markdown","metadata":{"id":"0zLxXGIQUuRt"},"source":["## Leitura dos arquivos .gff3\n","Vamos selecionar todas as linhas em que a Strand é positiva:"]},{"cell_type":"code","execution_count":35,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":424},"executionInfo":{"elapsed":1760,"status":"ok","timestamp":1732152630979,"user":{"displayName":"Fernando .Random. Henrique Paes Generich","userId":"13317897141628915437"},"user_tz":180},"id":"Luw5GIDHO4YT","outputId":"a0d388f1-89e8-4bf1-f072-3fbf4accee2d"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chr</th>\n","      <th>Class</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>7</td>\n","      <td>Class II subclass 2/Helitron/Helitron</td>\n","      <td>82856122</td>\n","      <td>82857584</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>1</td>\n","      <td>Class II subclass 2/Helitron/Helitron</td>\n","      <td>239302471</td>\n","      <td>239302834</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>8</td>\n","      <td>Class II subclass 2/Helitron/Helitron</td>\n","      <td>2261318</td>\n","      <td>2261604</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>5</td>\n","      <td>Class II subclass 2/Helitron/Helitron</td>\n","      <td>124665404</td>\n","      <td>124665608</td>\n","      <td>+</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>9</td>\n","      <td>Class II subclass 2/Helitron/Helitron</td>\n","      <td>77998293</td>\n","      <td>78002783</td>\n","      <td>+</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["  Chr                                  Class      Start        End Strand\n","0   7  Class II subclass 2/Helitron/Helitron   82856122   82857584      +\n","1   1  Class II subclass 2/Helitron/Helitron  239302471  239302834      +\n","2   8  Class II subclass 2/Helitron/Helitron    2261318    2261604      +\n","3   5  Class II subclass 2/Helitron/Helitron  124665404  124665608      +\n","4   9  Class II subclass 2/Helitron/Helitron   77998293   78002783      +"]},"execution_count":35,"metadata":{},"output_type":"execute_result"}],"source":["filenames = [\n","  'TEAnnotationFinal_Helitron.gff3',\n","  'TEAnnotationFinal_LINE.gff3',\n","  'TEAnnotationFinal_LTR.gff3',\n","  'TEAnnotationFinal_MITE.gff3',\n","  'TEAnnotationFinal_SINE.gff3',\n","  'TEAnnotationFinal_TIR.gff3'\n","]\n","\n","def process_file(filename: str) -> pd.DataFrame:\n","  df = pd.read_csv(f'data/{filename}', sep='\\t', header=None)\n","  df.drop(df[ df[6] != '+' ].index , inplace = True)\n","  df = df.drop( [1, 5, 7, 8], axis=1)\n","  return df\n","\n","te_df = pd.DataFrame()\n","\n","# Juntando os dados em um único DataFrame\n","for f in filenames:\n","  te_df = pd.concat([te_df, process_file(f)], ignore_index=True)\n","\n","# Renomeando colunas\n","te_df.rename(columns={0 : \"Chr\", 2: \"Class\", 3: \"Start\", 4: \"End\", 6: \"Strand\"}, inplace = True)\n","\n","# Seleciona Chr numérico apenas\n","te_df = te_df[te_df['Chr'].astype(str).str.isdigit()]\n","\n","te_df.head()"]},{"cell_type":"markdown","metadata":{"id":"RI8k8EHMDiZk"},"source":["## Obtendo as sequências de cromossomos do NCBI\n","No total, são 10 cromossomos"]},{"cell_type":"code","execution_count":5,"metadata":{"executionInfo":{"elapsed":66570,"status":"ok","timestamp":1732148635715,"user":{"displayName":"Fernando .Random. Henrique Paes Generich","userId":"13317897141628915437"},"user_tz":180},"id":"2uPDguA7KV2L"},"outputs":[{"name":"stderr","output_type":"stream","text":["  0%|          | 0/10 [00:00<?, ?it/s]"]},{"name":"stderr","output_type":"stream","text":["100%|██████████| 10/10 [09:06<00:00, 54.64s/it]\n"]}],"source":["Entrez.email = \"fernando_gene@usp.br\"\n","\n","ind = 618874\n","\n","allchromosomes = []\n","\n","for i in tqdm(range(10)):\n","  gen_bank_term = \"LR\" + str(ind+i) + \".1\"\n","\n","  handle = Entrez.esearch(db=\"nucleotide\", term=gen_bank_term, retmax=\"10\")\n","  rec_list = Entrez.read(handle)\n","  handle.close()\n","\n","  id_list = rec_list['IdList']\n","  handle = Entrez.efetch(db='nucleotide', id=id_list, rettype='fasta', retmode=\"text\")\n","  recs = list(SeqIO.parse(handle, 'fasta'))\n","  handle.close()\n","\n","  allchromosomes.append(recs[0])"]},{"cell_type":"code","execution_count":16,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":204},"executionInfo":{"elapsed":238,"status":"ok","timestamp":1732149291999,"user":{"displayName":"Fernando .Random. Henrique Paes Generich","userId":"13317897141628915437"},"user_tz":180},"id":"kXOToYfxw3h3","outputId":"8f7b6bad-4a8f-4d19-8371-a0a2d1747b52"},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Number</th>\n","      <th>Sequence</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>LR618874.1</td>\n","      <td>(T, C, A, T, G, G, C, T, A, T, T, T, T, C, A, ...</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>LR618875.1</td>\n","      <td>(T, T, C, T, C, A, T, T, A, T, A, T, C, C, T, ...</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>LR618876.1</td>\n","      <td>(C, C, T, A, A, A, C, C, C, T, A, A, A, C, C, ...</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>LR618877.1</td>\n","      <td>(C, C, T, A, A, A, C, C, C, T, A, A, A, C, C, ...</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>LR618878.1</td>\n","      <td>(C, T, A, A, A, C, C, T, A, A, A, C, A, T, C, ...</td>\n","    </tr>\n","    <tr>\n","      <th>5</th>\n","      <td>LR618879.1</td>\n","      <td>(A, A, A, A, C, C, C, T, A, A, A, C, C, C, T, ...</td>\n","    </tr>\n","    <tr>\n","      <th>6</th>\n","      <td>LR618880.1</td>\n","      <td>(C, T, A, A, A, A, C, C, C, T, A, A, A, C, C, ...</td>\n","    </tr>\n","    <tr>\n","      <th>7</th>\n","      <td>LR618881.1</td>\n","      <td>(A, A, A, C, C, T, A, A, A, C, C, C, T, A, A, ...</td>\n","    </tr>\n","    <tr>\n","      <th>8</th>\n","      <td>LR618882.1</td>\n","      <td>(G, T, C, G, C, T, C, A, T, G, G, C, T, A, T, ...</td>\n","    </tr>\n","    <tr>\n","      <th>9</th>\n","      <td>LR618883.1</td>\n","      <td>(T, A, A, A, C, C, C, T, A, A, A, C, C, C, T, ...</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["       Number                                           Sequence\n","0  LR618874.1  (T, C, A, T, G, G, C, T, A, T, T, T, T, C, A, ...\n","1  LR618875.1  (T, T, C, T, C, A, T, T, A, T, A, T, C, C, T, ...\n","2  LR618876.1  (C, C, T, A, A, A, C, C, C, T, A, A, A, C, C, ...\n","3  LR618877.1  (C, C, T, A, A, A, C, C, C, T, A, A, A, C, C, ...\n","4  LR618878.1  (C, T, A, A, A, C, C, T, A, A, A, C, A, T, C, ...\n","5  LR618879.1  (A, A, A, A, C, C, C, T, A, A, A, C, C, C, T, ...\n","6  LR618880.1  (C, T, A, A, A, A, C, C, C, T, A, A, A, C, C, ...\n","7  LR618881.1  (A, A, A, C, C, T, A, A, A, C, C, C, T, A, A, ...\n","8  LR618882.1  (G, T, C, G, C, T, C, A, T, G, G, C, T, A, T, ...\n","9  LR618883.1  (T, A, A, A, C, C, C, T, A, A, A, C, C, C, T, ..."]},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":["chromosome_dict = {\n","  'Number': [],\n","  'Sequence': []\n","}\n","\n","for i in allchromosomes:\n","  chromosome_dict['Number'].append(i.id)\n","  chromosome_dict['Sequence'].append(i.seq)\n","\n","chromosome_df = pd.DataFrame(chromosome_dict)\n","\n","chromosome_df"]},{"cell_type":"markdown","metadata":{},"source":["## Efetuando um Join entre os DataFrames "]},{"cell_type":"code","execution_count":37,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":193},"collapsed":true,"executionInfo":{"elapsed":5,"status":"error","timestamp":1732152655967,"user":{"displayName":"Fernando .Random. Henrique Paes Generich","userId":"13317897141628915437"},"user_tz":180},"id":"t6EaZnkEJftU","outputId":"7c6482e5-9a6c-4d6c-ce9c-93f38ebf5708"},"outputs":[],"source":["def get_sequence(chromosome: int, start: int, end: int) -> Seq.Seq:\n","  return chromosome_df.Sequence[chromosome-1][start:end+1]\n","\n","te_df['Sequence'] = te_df.apply(lambda x: get_sequence(int(x.Chr), int(x.Start), int(x.End)), axis=1)"]},{"cell_type":"markdown","metadata":{},"source":["Deleta linhas que possuem NaN"]},{"cell_type":"code","execution_count":10,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Chr</th>\n","      <th>Class</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Strand</th>\n","      <th>Sequence</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>7</td>\n","      <td>Class II subclass 2/Helitron/Helitron</td>\n","      <td>82856122</td>\n","      <td>82857584</td>\n","      <td>+</td>\n","      <td>AGCTTCGTCACCAGCTTTGCTCCGACCACCCTTTGTCCATACTAAC...</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>1</td>\n","      <td>Class II subclass 2/Helitron/Helitron</td>\n","      <td>239302471</td>\n","      <td>239302834</td>\n","      <td>+</td>\n","      <td>TCAGGGTTGCTTCTTGGCGAAGACAGGGCCTCGGGCGAGCCAGAAA...</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>8</td>\n","      <td>Class II subclass 2/Helitron/Helitron</td>\n","      <td>2261318</td>\n","      <td>2261604</td>\n","      <td>+</td>\n","      <td>CGCCCAAGCAGACGGTCACCATCAGCGAAGACCTCACTTCGCATGA...</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>5</td>\n","      <td>Class II subclass 2/Helitron/Helitron</td>\n","      <td>124665404</td>\n","      <td>124665608</td>\n","      <td>+</td>\n","      <td>ATGCCAAGTCGTGTCAAACGACTTAGGGTAGGGGTCAACTTTCTCC...</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>9</td>\n","      <td>Class II subclass 2/Helitron/Helitron</td>\n","      <td>77998293</td>\n","      <td>78002783</td>\n","      <td>+</td>\n","      <td>TTAGGTTATTTATATACTAGTTTATGTTGATGATATAATCATCACT...</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["   Chr                                  Class      Start        End Strand  \\\n","0    7  Class II subclass 2/Helitron/Helitron   82856122   82857584      +   \n","1    1  Class II subclass 2/Helitron/Helitron  239302471  239302834      +   \n","2    8  Class II subclass 2/Helitron/Helitron    2261318    2261604      +   \n","3    5  Class II subclass 2/Helitron/Helitron  124665404  124665608      +   \n","4    9  Class II subclass 2/Helitron/Helitron   77998293   78002783      +   \n","\n","                                            Sequence  \n","0  AGCTTCGTCACCAGCTTTGCTCCGACCACCCTTTGTCCATACTAAC...  \n","1  TCAGGGTTGCTTCTTGGCGAAGACAGGGCCTCGGGCGAGCCAGAAA...  \n","2  CGCCCAAGCAGACGGTCACCATCAGCGAAGACCTCACTTCGCATGA...  \n","3  ATGCCAAGTCGTGTCAAACGACTTAGGGTAGGGGTCAACTTTCTCC...  \n","4  TTAGGTTATTTATATACTAGTTTATGTTGATGATATAATCATCACT...  "]},"execution_count":10,"metadata":{},"output_type":"execute_result"}],"source":["te_df = te_df.dropna()\n","te_df.head()"]},{"cell_type":"markdown","metadata":{},"source":["Salva o Dataframe como `.csv`"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["te_df.to_csv('data/transposable_elements.csv')"]},{"cell_type":"markdown","metadata":{},"source":["Leitura do arquivo csv"]},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[],"source":["te_df = pd.read_csv('data/tranposable_elements.csv', index_col=0)"]},{"cell_type":"markdown","metadata":{},"source":["Vamos criar arquivos fasta para inserir no MathFeature"]},{"cell_type":"code","execution_count":44,"metadata":{},"outputs":[],"source":["sequences = []\n","for index, sequence in list(zip(te_df.index, te_df['Sequence'])):\n","    sequences.append(SeqRecord.SeqRecord(Seq.Seq(sequence), id=str(index)))"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]}],"metadata":{"colab":{"authorship_tag":"ABX9TyOfY4uNH4+85oO6s9it0PvH","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.12.6"}},"nbformat":4,"nbformat_minor":0}
